crates/spirv-std/src/image/params.rs

```diff
@@ -2,60 +2,85 @@ use super::{Arrayed, Dimensionality, ImageFormat};
-impl SampleType<{ ImageFormat::Unknown as u32 }> for u32 {}
-impl SampleType<{ ImageFormat::Unknown as u32 }> for u64 {}
-impl SampleType<{ ImageFormat::Unknown as u32 }> for f32 {}
-impl SampleType<{ ImageFormat::Unknown as u32 }> for f64 {}
-impl SampleType<{ ImageFormat::Rgba32f as u32 }> for f32 {}
-impl SampleType<{ ImageFormat::Rgba16f as u32 }> for f32 {}
-impl SampleType<{ ImageFormat::R32f as u32 }> for f32 {}
-impl SampleType<{ ImageFormat::Rgba8 as u32 }> for f32 {}
-impl SampleType<{ ImageFormat::Rgba8Snorm as u32 }> for f32 {}
-impl SampleType<{ ImageFormat::Rg32f as u32 }> for f32 {}
-impl SampleType<{ ImageFormat::Rg16f as u32 }> for f32 {}
-impl SampleType<{ ImageFormat::R11fG11fB10f as u32 }> for f32 {}
-impl SampleType<{ ImageFormat::R16f as u32 }> for f32 {}
-impl SampleType<{ ImageFormat::Rgba16 as u32 }> for f32 {}
-impl SampleType<{ ImageFormat::Rgb10A2 as u32 }> for f32 {}
-impl SampleType<{ ImageFormat::Rg16 as u32 }> for f32 {}
-impl SampleType<{ ImageFormat::Rg8 as u32 }> for f32 {}
-impl SampleType<{ ImageFormat::R16 as u32 }> for f32 {}
-impl SampleType<{ ImageFormat::R8 as u32 }> for f32 {}
-impl SampleType<{ ImageFormat::Rgba16Snorm as u32 }> for f32 {}
-impl SampleType<{ ImageFormat::Rg16Snorm as u32 }> for f32 {}
-impl SampleType<{ ImageFormat::Rg8Snorm as u32 }> for f32 {}
-impl SampleType<{ ImageFormat::R16Snorm as u32 }> for f32 {}
-impl SampleType<{ ImageFormat::R8Snorm as u32 }> for f32 {}
-impl SampleType<{ ImageFormat::Rgba32i as u32 }> for i32 {}
-impl SampleType<{ ImageFormat::Rgba16i as u32 }> for i32 {}
-impl SampleType<{ ImageFormat::Rgba8i as u32 }> for i32 {}
-impl SampleType<{ ImageFormat::R32i as u32 }> for i32 {}
-impl SampleType<{ ImageFormat::Rg32i as u32 }> for i32 {}
-impl SampleType<{ ImageFormat::Rg16i as u32 }> for i32 {}
-impl SampleType<{ ImageFormat::Rg8i as u32 }> for i32 {}
-impl SampleType<{ ImageFormat::R16i as u32 }> for i32 {}
-impl SampleType<{ ImageFormat::R8i as u32 }> for i32 {}
-impl SampleType<{ ImageFormat::Rgba32ui as u32 }> for u32 {}
-impl SampleType<{ ImageFormat::Rgba16ui as u32 }> for u32 {}
-impl SampleType<{ ImageFormat::Rgba8ui as u32 }> for u32 {}
-impl SampleType<{ ImageFormat::R32ui as u32 }> for u32 {}
-impl SampleType<{ ImageFormat::Rgb10A2ui as u32 }> for u32 {}
-impl SampleType<{ ImageFormat::Rg32ui as u32 }> for u32 {}
-impl SampleType<{ ImageFormat::Rg16ui as u32 }> for u32 {}
-impl SampleType<{ ImageFormat::Rg8ui as u32 }> for u32 {}
-impl SampleType<{ ImageFormat::R16ui as u32 }> for u32 {}
-impl SampleType<{ ImageFormat::R8ui as u32 }> for u32 {}
-impl SampleType<{ ImageFormat::R64ui as u32 }> for u64 {}
-impl SampleType<{ ImageFormat::R64i as u32 }> for i64 {}
+    /// A 3-component vector of this sample type
+    type Vec3: Default;
+
+    /// A 4-component vector of this sample type
+    type Vec4: Default;
+}
+
+/// Helper macro to implement `SampleType` of various formats for various scalar types.
+macro_rules! sampletype_impls {
```

Comment:
Nit: `FooBar` typically turns into `foo_bar` when switching to snake case (i.e. `sample_type_impls`).